cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/lineage/LineageOperationsProcessor.java

```diff
@@ -296,9 +314,19 @@ private List<InputField> createInputFields(List<String> fields, String currentSt
         // first
         String stage = parentsIterator.previous();
         // check if the current field is output by any one of operation created by current stage
-        String origin = stageOutputsWithOrigins.get(stage).get(actualField);
-        if (origin != null) {
-          inputFields.put(field, InputField.of(origin, actualField));
+        List<Origin> possibleOrigins = stageOutputsWithOrigins.get(stage).get(actualField);
+        if (possibleOrigins != null) {
+          // if there is only one origin, update the input field. In this case, tag will be null
+          // if there is more than one origin, we need to match the tag to find the correct input
+          if (possibleOrigins.size() == 1) {
+            inputFields.put(field, InputField.of(possibleOrigins.get(0).getStageOperation(), actualField));
+          } else {
+            for (Origin possibleOrigin: possibleOrigins) {
+              if (tag != null && tag.equals(possibleOrigin.getTag())) {
+                inputFields.put(field, InputField.of(possibleOrigin.getStageOperation(), actualField));
```

Comment:
Discussed offline.
tag will not be null here because while updating the origin, if there is no tag there will be only origin
If there are multiple origins, tag will not be null. Added comments here and where we update the origin